### PR TITLE
s3ext: add a standalone config checking app [v2]

### PR DESCRIPTION
**Pull request by adam8157**
Usage: s3chkcfg -c "s3://endpoint/bucket/prefix config=path_to_config_file", to check the configuration.
       s3chkcfg -d "s3://endpoint/bucket/prefix config=path_to_config_file", to download and output to stdout.
       s3chkcfg -t, to show the config template.
       s3chkcfg -h, to show this help.

Also did refactoring to reuse functions common with s3ext module.

### CODE REVIEW

File: gpAux/extensions/gps3ext/src/s3chkcfg.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+#include "s3chkcfg.h"
+
+volatile bool QueryCancelPending = false;
+
+int main(int argc, char *argv[]) {
+    int opt = 0;
+    int ret = 0;
+
+    s3ext_logtype = STDERR_LOG;
```
Comment (lij55) [inline gpAux/extensions/gps3ext/src/s3chkcfg.cpp]: let user control these level

File: gpAux/extensions/gps3ext/src/s3chkcfg.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    bool ret = false;
+
+    ret = InitConfig(config, "default");
+    s3ext_logtype = STDERR_LOG;
+    s3ext_loglevel = EXT_ERROR;
+
+    return ret;
+}
+
+ListBucketResult *list_bucket(S3Reader *wrapper) {
```
Comment (lij55) [inline gpAux/extensions/gps3ext/src/s3chkcfg.cpp]: Use same function as tested

File: gpAux/extensions/gps3ext/src/s3chkcfg.cpp [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        return false;
+    }
+
+    S3Reader *wrapper = NULL;
+    ListBucketResult *result = NULL;
+    bool ret = false;
+
+    if (!read_config(config_path)) {
+        goto FAIL;
+    }
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/s3chkcfg.cpp]: Able to reuse code in reader_init().

File: gpAux/extensions/gps3ext/src/s3chkcfg.cpp [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    thread_setup();
+
+    char *data_buf = (char *)malloc(BUF_SIZE);
+    if (!data_buf) {
+        goto FAIL;
+    }
+
+    s3ext_logtype = REMOTE_LOG;
+    s3ext_loglevel = EXT_ERROR;
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/s3chkcfg.cpp]: hardcode logtype and level?

**Comment by yydzero** [comment]
For error messages, suggest to include as much information as possible, eg: 'could not find config file', change to sth. like "could not find config file: '/path/to/config.file'"

File: gpAux/extensions/gps3ext/src/s3chkcfg.cpp [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    return r;
+}
+
+uint8_t print_contents(ListBucketResult *r) {
+    char urlbuf[256];
+    uint8_t count = 0;
+    vector<BucketContent *>::iterator i;
+
+    for (i = r->contents.begin(); i != r->contents.end(); i++) {
+        if (count > 8) {
```
Comment (lij55) [inline gpAux/extensions/gps3ext/src/s3chkcfg.cpp]: limitation is needless

**Comment by adam8157** [comment]
Added a patch, still hardcoded the logtype to stderr, and not to reuse reader_init() since it has not many common codes.

File: gpAux/extensions/gps3ext/src/s3conf.cpp [hunk 6]
```diff
@@ -76,9 +76,9 @@ bool InitConfig(const string& conf_path,
         s3cfg = new Config(conf_path);
         if (!s3cfg || !s3cfg->Handle()) {
 #ifndef DEBUG_S3
-            write_log("Failed to parse config file\n");
+            write_log("Failed to parse config file, or it doesn't exist\n");
 #else
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/s3conf.cpp]: could we print filename also?

**Comment by adam8157** [comment]
Have held a group review today, issues are patched now.

**Pull request merged**